safegraph_py_functions/safegraph_py_functions.py

```diff
@@ -186,6 +196,93 @@ def explode_json_array(df_, array_column = 'visits_by_day', value_col_name=None,
 
 ### ------------------------------------------ END JSON SECTION--------------------------------------------------------
 
+### ------------------------------------------ JSON FAST SECTION--------------------------------------------------------
+def process_unpack_json(df, json_column, index_name, key_col_name, value_col_name):
+    df[json_column + '_dict'] = df[json_column].apply(lambda x: json.loads(x) if type(x) == str else x) # json.loads breaks if the column is empty, so this function is a work around
+    all_sgpid_cbg_data = []  # each cbg data point will be one element in this list
+    if index_name == None:
+      for index, row in df.iterrows():
+          this_sgpid_cbg_data = [{'orig_index': index, key_col_name: key, value_col_name: value} for key, value in
+                                row[json_column + '_dict'].items()]
+          all_sgpid_cbg_data = all_sgpid_cbg_data + this_sgpid_cbg_data
+    else:
+      for index, row in df.iterrows():
+        temp = row[index_name]
+        this_sgpid_cbg_data = [{'orig_index': index, index_name:temp, key_col_name: key, value_col_name: value} for key, value in
+                               row[json_column + '_dict'].items()]
+        all_sgpid_cbg_data = all_sgpid_cbg_data + this_sgpid_cbg_data
+    
+    del df # free memory 
+    all_sgpid_cbg_data = pd.DataFrame(all_sgpid_cbg_data)
+    all_sgpid_cbg_data.set_index('orig_index', inplace=True)
+    return all_sgpid_cbg_data
+
+# index_name if you want your index (such as CBG) to be it's own column, then provide this 
+def unpack_json_fast(df, json_column = 'visitor_home_cbgs', index_name = None, key_col_name = None, value_col_name = None, chunk_n = 1000):
+    # figure out column names, defaulting to 'column name' + _attribute
+    if key_col_name == None:
+      key_col_name = json_column + '_key'
+    if value_col_name == None:
+      value_col_name = json_column + '_value'
+    if (df.index.unique().shape[0] < df.shape[0]):
+      raise ("ERROR -- non-unique index found")
+    
+    chunks_list = [df[i:i+chunk_n] for i in range(0,df.shape[0],chunk_n)] # splitting helps with memory, but it also has an impact on run time. 
+    del df # free memory (Dont think we can avoid the DF being loaded in twice for a moment)
+
+    function = partial(process_unpack_json, json_column=json_column, index_name= index_name, key_col_name= key_col_name, value_col_name= value_col_name)
```

Comment:
@bpblakely  can we name this something less generic, so taht it is easy to search for in the code? the word `function` already appears nearly 40 times in this file. thanks

Reply:
Sure we can rename ```function``` and yeah I did misunderstand what you meant with 2 & 4. I'll look into that further.

3. I updated both the normal and fast function by removing the redundant checking.

5. I'll generate some graphs to double check the memory usage, but for now I'll just omit the deletes.

Will update the pull request when I figure out a good way to do 2 & 4

Reply:
I've got a version of the code working where multithreading uses the normal function and no longer needs to use a separate function. The only downside is if you do it this way your mulithreading performance gets impacted (by a few seconds in medium size data sets), since you can no longer control the things that only need to get checked once (like initializing key values), so it gets checked every single time you run a partition.